src/ast/passes/codegen_llvm.cpp

```diff
@@ -2465,6 +2472,72 @@ void CodegenLLVM::generateProbe(Probe &probe,
         func_type, section_name, current_attach_point_->address, index);
 }
 
+void CodegenLLVM::visit(Subprog &subprog)
+{
+  std::vector<llvm::Type *> arg_types;
+  std::transform(subprog.args->begin(),
+                 subprog.args->end(),
+                 std::back_inserter(arg_types),
+                 [this](SubprogArg *arg) { return b_.GetType(arg->type); });
+  FunctionType *func_type = FunctionType::get(b_.GetType(subprog.type),
+                                              arg_types,
+                                              0);
+
+  Function *func = Function::Create(
+      func_type, Function::InternalLinkage, subprog.name(), module_.get());
+  BasicBlock *entry = BasicBlock::Create(module_->getContext(), "entry", func);
+  b_.SetInsertPoint(entry);
+
+  variables_.clear();
+  ctx_ = nullptr;
+  inside_subprog_ = true;
+
+  int arg_index = 0;
+  for (SubprogArg *arg : *subprog.args)
+  {
+    auto alloca = b_.CreateAllocaBPF(b_.GetType(arg->type), arg->name());
+    b_.CreateStore(func->getArg(arg_index), alloca);
+    variables_.insert({ arg->name(), alloca });
+    ++arg_index;
+  }
+
+  for (Statement *stmt : *subprog.stmts)
+  {
+    auto scoped_del = accept(stmt);
+  }
+  createRet();
+
+  if (!subprog.type.IsVoidTy())
+  {
+    // Check that every reachable return is non-dummy (for dummy returns, see
+    // comment in CodegenLLVM::createRet)
+    // Note that semantic analyser checks that we use a correct return type
+    auto start_bb = &func->getEntryBlock();
+    std::deque<llvm::BasicBlock *> bfs_queue{ start_bb };
+    std::unordered_set<llvm::BasicBlock *> bfs_visited{ start_bb };
+    while (!bfs_queue.empty())
+    {
+      auto bb = bfs_queue.back();
+      bfs_queue.pop_back();
+      auto ret = llvm::dyn_cast<llvm::ReturnInst>(bb->getTerminator());
+      if (ret && ret->getMetadata("ret_dummy"))
+      {
+        LOG(ERROR, subprog.loc, err_) << "Not all paths in function "
+                                      << subprog.name() << " return a value";
+      }
+      auto term = bb->getTerminator();
+      for (unsigned i = 0, e = term->getNumSuccessors(); i != e; i++)
+      {
+        auto succ = term->getSuccessor(i);
+        if (bfs_visited.find(succ) != bfs_visited.end())
+          continue;
+        bfs_visited.insert(succ);
+        bfs_queue.push_front(succ);
+      }
+    }
```

Comment:
This still applies, right?

Reply:
Ah yes, I somehow missed that you are suggesting to run unreachable code elimination. The codegen already generates dummy returns for probes, they are just simpler, since the return value of probes is either void or some kind of integer, so there is no need for an empty alloca.

On the other hand, running dead code elimination here might slightly impact the tests (e.g. the pass might fail on invalid LLVM IR or errors in unreachable code won't be caught in verification). I will try to put the pass there and see what it does.

Reply:
I tried running `UnreachableBlockElimPass` at the end of `void CodegenLLVM::visit(Subprog &subprog)` and it looks good. I think it might make sense to do that in a separate PR together with the same pass for probes rather than here, as I wrote above, this is not an issue specific to user-defined functions, it is just the most visible there.